src/main/scala/io/qbeast/spark/index/OTreeDataAnalyzer.scala

```diff
@@ -160,8 +142,171 @@ object DoublePassOTreeDataAnalyzer extends OTreeDataAnalyzer with Serializable {
         })
+
     }
 
+  /**
+   * Extract tree size and domain from a given local tree.
+   * The domain of a missing cube is computed from its immediate parent node,
+   * in a top-down fashion.
+   * @param cube CubeId whose tree size and domain are of our interest
+   * @param localTree LocalTree from which to extract tree size and domain
+   * @param treeSizeDomain existing tree sizes and domains
+   * @param isReplication whether the current process is an optimization process
+   * @return
+   */
+  private[index] def computeLocalTreeSizeAndDomain(
+      cube: CubeId,
+      localTree: LocalTree,
+      treeSizeDomain: Map[CubeId, TreeSizeAndDomain],
+      isReplication: Boolean): TreeSizeAndDomain = {
+    localTree.get(cube) match {
+      case Some(info) =>
+        // Cube exists in the local tree, compute domain from tree size.
+        TreeSizeAndDomain(info.treeSize, info.treeSize / (1d - info.parentWeight))
+      case None =>
+        // Cube missing from the local tree, compute domain from its parent domain.
+        // Root should always exists in all local trees, and it's never present
+        // during replication.
+        val parent = cube.parent.get
+        val (parentTreeSize, parentDomain) =
+          if (isReplication && !treeSizeDomain.contains(parent)) {
+            (0d, 0d)
+          } else {
+            val ptsd = treeSizeDomain(parent)
+            (ptsd.treeSize, ptsd.domain)
+          }
+        val siblings = parent.children.filter(localTree.contains)
+        val domain = if (siblings.isEmpty) {
+          // Parent cube has no child cubes. Consider its domain is uniquely
+          // composed of that of the missing cube.
+          // Extract payload size from parent value to avoid overestimating missing cube domain.
+          parentDomain - parentTreeSize
+        } else {
+          // Assume parent domain distribution follows that of the its child node tree sizes,
+          // and the missing cube tree size is 1.
+          val siblingSize = siblings.map(localTree(_).treeSize).sum
+          val f = 1d / (siblingSize + 1d)
+          f * parentDomain
+        }
+        TreeSizeAndDomain(0d, domain)
+    }
+  }
+
+  /**
+   * Compute cube and domain from all LocalTrees by broadcasting
+   * the unique cubes from all partitions.
+   * @param localTrees the collection of LocalTrees from all partitions
+   * @param isReplication whether the current process is a replication process
+   * @param dimensionCount the number of indexing columns
+   * @param desiredCubeSize the desired cube size for the index
+   * @return
+   */
+  private[index] def computeGlobalDomain(
+      localTrees: Dataset[LocalTree],
+      isReplication: Boolean,
+      dimensionCount: Int,
+      desiredCubeSize: Int): Seq[(CubeId, Double)] = {
+    val spark = SparkSession.active
+    import spark.implicits._
+
+    // Unique cubes among all local trees
+    val globalCubes = spark.sparkContext.broadcast(
```

Comment:
A collect, to do a broadcast and then use it in another dataframe doesn't seem very efficient. Can we do the same with a JOIN (e.g. an outer join)?